Support app-specific web apps permissions for web apps users

orangejenny opened the pull request:
## Product Description

### Overview
https://dimagi.atlassian.net/browse/USH-4289

This work overlaps somewhat with the Login As changes @MartinRiese is doing in https://github.com/dimagi/commcare-hq/pull/34351 This description is written assuming that those changes are complete. I'm planning to wait on QAing this PR until that work is complete.

### Old behavior
There's a single permission in roles & permissions that controls access to web apps.

![Screenshot 2024-04-11 at 2 17 51 PM](https://github.com/dimagi/commcare-hq/assets/1486591/f988213c-3d83-42bd-951b-e7988ab38e0f)

This permission is checked for the user who is logged into HQ. When Login As is used, permissions of the Login As user are ignored. Also, the web apps permission is not checked if the user logged into HQ is a mobile user. The default role for mobile workers doesn't even include web apps access ([see here](https://github.com/dimagi/commcare-hq/blob/49117bc516e1f5a48fc686823bbafa34bf27e23f/corehq/apps/users/role_utils.py#L29-L31)). All of this is controlled by the [@require_cloudcare_access](https://github.com/dimagi/commcare-hq/blob/49117bc516e1f5a48fc686823bbafa34bf27e23f/corehq/apps/cloudcare/decorators.py#L15-L50) decorator, which is applied to cloudcare views.

In addition, access can be granted to specific applications in web apps using a bespoke set of models and UI:

![Screenshot 2024-04-11 at 2 24 22 PM](https://github.com/dimagi/commcare-hq/assets/1486591/0dccc8d3-908e-4fdb-a883-38fd94fbdf92)

This is based on mapping applications to user groups. User groups are only available for mobile users. Therefore, app-specific permissions are only available for mobile users.

### Impetus for change

As part of the work to move web apps projects away from depending on linked mobile users, we need to support app-specific permissions for web users.

There's an alternative solution where we add support for web users to user groups. This is discussed in the ticket and linked documents, but in short, it was rejected in favor of integrating app-specific web apps permissions with our general roles & permissions framework.

### New behavior

I'm clubbing a couple of related changes into this PR in the interest of getting holistic review. I'm happy to split things up if reviewers would find that helpful or believe it would be safer.

1. Add app-specific permissions to roles & permissions (first few commits)

This is a straightforward addition to our existing roles & permissions framework.
![Screenshot 2024-04-11 at 2 35 13 PM](https://github.com/dimagi/commcare-hq/assets/1486591/a9ea9f55-bba3-4856-a048-d7888599c1c3)

2. Add feature flag to hide old group-based behavior (next few commits)

Having two ways to assign app-specific permissions is confusing. Moving from group-based to role-based permissions requires human intervention. As a compromise, I'm adding a feature flag, marked deprecated, to gate the old UI. Migration `0059` turns this flag on for all domains that currently have an `ApplicationAccess` object that's restricting app access. I added a warning to the top of the old UI, but I'm not changing behavior at all. ApplicationAccess restrictions will continue to be applied as they are today, in addition to the new app-specific permissions.

~~3. Update default mobile worker role to access web apps, and remove mobile worker check (last few commits)~~

~~This is the broadest/scariest change. Pushing users to use roles to manage app-specific permissions, for both web and mobile workers, makes it more glaringly apparent that the existing web apps permission doesn't apply to mobile workers. To reduce ambiguity, I'm updating `require_cloudcare_access` to check the permission for both web and mobile users. To avoid a change in behavior, migration `0060` updates all of the default mobile worker roles to turn on web apps access.~~

How do these changes combine? I'm continuing to decorate cloudcare views with a permissions check on the HQ user, so if you log into HQ and don't have any cloudcare permissions, you always 403, no opportunity to login as someone else. If you can log into HQ and then use login as, that user's permissions are used to determine which apps you see. This does mean that you can log into HQ as a user who has access only to app A, and can login as a user who has access to app B.

Summary of behavior:
_This table has been updated to reflect that part 3 of "new behavior" above has been reverted._
| HQ user `access_web_apps` | Login As user `access_web_apps` | Behavior |
| ------------- | ------------- | ------------- |
| False | no login as | 403 |
| app-specific | no login as | display apps HQ user has access to |
| | False | all apps displayed |
|  | app-specific | display apps **login as user** has access to |
|  | True | all apps displayed |
| True | no login as | all apps displayed |
| | False | all apps displayed |
|  | app-specific | display apps **login as user** has access to |
|  | True | all apps displayed |

In **addition** to this, any legacy `ApplicationAccess` restrictions will continue to be applied, for backwards compatibility.

## Feature Flag
Introduces WEB_APPS_PERMISSIONS_VIA_GROUPS, but most of this behavior is not gated.

## Safety Assurance

### Safety story
Permissions are always sensitive, and this does migrations ~~and changes default-type behavior~~, so it feels medium-high risk.

### Automated test coverage

There are tests for roles and permissions.

### QA Plan

https://dimagi.atlassian.net/browse/QA-6352

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [ ] The migrations in this code can be safely applied first independently of the code

No? The migrations are integrated with the code. I'm not terribly concerned that someone will add an `ApplicationAccess` object during deploy, after migration `0059` runs but before the new code that restricts the config UI based on that feature flag goes live.

I could pull the application access changes (the new feature flag, migration `0059`, and the changes that restrict the old config UI) into a separate PR, or even two separate PRs (one for the flag + migration, a second for the UI restrictions). I'm not going to do that unless anyone else is concerned about the deploy window, because I think the risk here is low.

~~For the `0060` migration for the default mobile worker role, the current behavior is at best ambiguous and arguably incorrect, since a mobile worker with a role that doesn't have access to web apps still gets access to web apps. That makes me not concerned about default mobile worker roles getting updated during deploy.~~

### Rollback instructions

- [ ] This PR can be reverted after deploy with no further considerations

Rolling back will not be possible as soon as people start adding app-specific permissions to the database, as the wrap function will start to fail.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

orangejenny committed:
Added web_apps_list to model

orangejenny committed:
Made list of web apps available to javascript

orangejenny committed:
Removed unused initial page data

The rest of this was removed in https://github.com/dimagi/commcare-hq/pull/29767/

orangejenny committed:
Added apps list to UI

orangejenny committed:
Pulled linked domain privileges from privileges module instead of initial page data

Privileges are already available on the front end.

orangejenny committed:
Extracted template for the access all/none/selected UI

orangejenny committed:
Updated web apps to filter out apps that user doesn't have permissions for

orangejenny committed:
Added flag to gate old behavior

orangejenny committed:
Moved legacy UI behind feature flag

orangejenny committed:
Stopped allowing all mobile users access to web apps

esoergel commented on corehq/apps/users/static/users/js/roles.js:
```diff
@@ -267,22 +279,6 @@ hqDefine('users/js/roles',[
                         allowCheckboxPermission: null,
                     },
-                    {
-                        showOption: root.webAppsPrivilege,
```
I'm guessing this was for projects that have plans without access to web apps (is that Community?).  Removing this restriction will mean that those projects will now be able to configure web apps permissions even when not applicable.  Probably not a big deal, just wanted to check.

esoergel replied:
nevermind, I see it was functionally replaced with
```            {% if request|request_has_privilege:"CLOUDCARE" %}```
below

esoergel commented on corehq/apps/users/templates/users/partials/edit_role_modal.html:
```diff
@@ -206,6 +206,52 @@ <h4 class="modal-title" data-bind="text: modalTitle"></h4>
+            {% if request|request_has_privilege:"CLOUDCARE" %}
+            <legend>
+              {% trans "Web Apps" %}
+            </legend>
```
Did this look like a decent target to extract into a template?  Not sure how much it has in common with the other parameterized permissions

esoergel commented on corehq/apps/users/views/__init__.py:
```diff
@@ -510,20 +510,6 @@ def can_restrict_access_by_location(self):
 
-    @property
-    @memoized
-    def release_management_privilege(self):
```
nice cleanup commit

esoergel commented on corehq/apps/users/templates/users/partials/edit_role_modal.html:
```diff
@@ -556,3 +478,43 @@ <h4 class="modal-title" data-bind="text: modalTitle"></h4>
   </div>
 </div>
+
+<script type="text/html" id="permission_all_selected_none">
```
lol, you're always a step ahead

orangejenny replied:
😀

esoergel commented on corehq/apps/users/migrations/0059_add_web_apps_groups_permission_flag.py:
```diff
@@ -0,0 +1,22 @@
+
+@skip_on_fresh_install
+def enable_flag(apps, schema_editor):
+    for access in ApplicationAccess.objects.filter(restrict=True):
+        WEB_APPS_PERMISSIONS_VIA_GROUPS.set(access.domain, enabled=True, namespace=NAMESPACE_DOMAIN)
```
are you anticipating this being fast enough to run during the deploy?

orangejenny replied:
yeah, with only a couple hundred `ApplicationAccess` objects on prod this should be fast

esoergel commented on corehq/toggles/__init__.py:
```diff
@@ -944,6 +944,13 @@ def _ensure_valid_randomness(randomness):
     """
 )
 
+WEB_APPS_PERMISSIONS_VIA_GROUPS = StaticToggle(
```
Might throw in a description pointing to the alternative, unless you did that on the page itself

esoergel commented on corehq/apps/cloudcare/templates/cloudcare/config.html:
```diff
@@ -11,6 +11,13 @@
+    {% blocktrans %}
+      This feature is deprecated. Access to specific web apps can now be managed via
+      <a target="_blank" href="{{ roles_url }}">Roles & Permissions</a>.
+    {% endblocktrans %}
```
lol, yup there it is

orangejenny replied:
It's a good point that the flag itself could be more descriptive, updated in 7aa89a67b64

esoergel commented on corehq/apps/users/migrations/0059_add_web_apps_groups_permission_flag.py:
```diff
@@ -0,0 +1,22 @@
+
+@skip_on_fresh_install
+def enable_flag(apps, schema_editor):
+    for access in ApplicationAccess.objects.filter(restrict=True):
```
Is this as targeted as we can safely do?  Like, would it make sense to check whether the domain has access to web apps or groups configured or some such?  I guess that could also be a clean-up pass done separately later

orangejenny replied:
I think it's pretty harmless, the view itself is still decorated to require the cloudcare privilege. The risks here are:
* Block access to this page for a domain that should have it: I think this is unlikely, although I could remove the `restrict=True` filter.
* Accidentally give access to a domain that doesn't need it: not a big deal, since right now all domains (with cloudcare privilege) can access this page

esoergel replied:
> Accidentally give access to a domain that doesn't need it: not a big deal, since right now all domains (with cloudcare privilege) can access this page
Yeah, I agree - this is the situation I was thinking of.  Fewer domains on this flag means that it's easier to eventually remove in the future, but I agree it's probably better to play it safe here and work on whittling down the list separately whenever we get around to it

esoergel commented on corehq/apps/users/migrations/0060_grant_web_apps_default_mobile_worker.py:
```diff
@@ -0,0 +1,25 @@
+
+
+@skip_on_fresh_install
+def _update_default_mobile_worker_role(apps, schema_editor):
```
I'm also a bit nervous about the access change associated with this.  There's always the possibility that there are mobile workers using roles other than the default _and_ using web apps.  I do think the current behavior is arguably a bug (if the role explicitly doesn't have web apps access...), so I think we have some wiggle room.  Still, is there anything we could do to de-risk this, or get a sense of how many users it might realistically impact?

orangejenny replied:
Point well taken. I'll do some digging on this and report back.

(You reviewed this a lot faster than I was expecting - thank you!)

orangejenny replied:
reverted this change, see longer comment below

esoergel reviewed (approved):
Exemplary PR, bravo!  Great description, commit breakdown, etc

orangejenny committed:
Added description for WEB_APPS_PERMISSIONS_VIA_GROUPS flag

snopoke commented on corehq/tabs/tabclasses.py:
```diff
@@ -1153,7 +1153,7 @@ def _is_viewable(self):
             has_privilege(self._request, privileges.CLOUDCARE)
             and self.domain
-            and (self.couch_user.can_access_web_apps() or self.couch_user.is_commcare_user())
+            and (self.couch_user.can_access_any_web_apps() or self.couch_user.is_commcare_user())
```
Why is the `self.couch_user.is_commcare_user()` check still necessary here?

orangejenny replied:
thanks - 27c0cc3fca7

snopoke reviewed (commented):
This is a great change, thanks! If a web user doesn't have access to any apps can they still get the Web Apps page and use login as?

MartinRiese commented:
Can't add anything useful. But it is good learning to read through the PR and Ethan's comments.

Robert-Costello commented:
> Can't add anything useful. But it is good learning to read through the PR and Ethan's comments.

Same here. Reading through this PR and the comments is especially helpful since I'm starting on adjacent permissions work for tableau access.

orangejenny committed:
Replaced _id check with copy_of

orangejenny commented on corehq/apps/cloudcare/views.py:
```diff
@@ -120,6 +120,7 @@ def get_web_apps_available_to_user(self, domain, user):
         ))
         apps = filter(None, apps)
         apps = filter(lambda app: app.get('cloudcare_enabled') or self.preview, apps)
+        apps = filter(lambda app: user.can_access_web_app(domain, app.get('_id')), apps)
```
Oops, yes. `origin_id` isn't available because these are docs not objects, but I can update to use the same logic. 9c9512926c8

orangejenny committed:
Removed leftover is_commcare_user check

esoergel commented on corehq/toggles/__init__.py:
```diff
@@ -946,9 +946,14 @@ def _ensure_valid_randomness(randomness):
+    description="""
+        This is a legacy feature that allows setting app-specific permissions in Web Apps based on user
+        groups. This functionality is now available via the roles and permissions page, where the
+        permission for accessing web apps supports specifying a list of apps.
```
:+1:

esoergel commented on corehq/tabs/tabclasses.py:
```diff
@@ -1153,7 +1153,7 @@ def _is_viewable(self):
             has_privilege(self._request, privileges.CLOUDCARE)
             and self.domain
-            and (self.couch_user.can_access_any_web_apps() or self.couch_user.is_commcare_user())
+            and self.couch_user.can_access_any_web_apps()
```
this feels really satisfying

orangejenny replied:
reverting this change, see longer comment below

millerdev reviewed (commented):
Looks good to me. Couple of small questions. The feature flag question is the only one that may be blocking, depending on the answer.

> ![image](https://github.com/dimagi/commcare-hq/assets/464726/87753adb-af54-4a93-9f46-46633448c5b2)

The _Save_ button and radio buttons are a bit crowded, to the point where I initially didn't see the top radio button. Is that typical spacing for those elements? Would it make sense to give the radio buttons a bit more breathing room?

> Add feature flag to hide old group-based behavior...
>
> ApplicationAccess restrictions will continue to be applied as they are today, in addition to the new app-specific permissions.

Not sure if I'm understanding this correctly. Could it be confusing if, say, a domain configured the new app-specific permissions while also having (pre-existing) conflicting ApplicationAccess restrictions, and had no way to edit the latter as a result of the feature flag being enabled? If yes, would it make sense to ignore ApplicationAccess restrictions if the feature flag is enabled and any new app-specific permissions have been configured?

orangejenny commented on corehq/apps/users/models.py:
```diff
@@ -1577,6 +1582,17 @@ def can_access_any_exports(self, domain=None):
+        if domain:
+            try:
+                role = self.get_role(domain)
+                return any(role.permissions.web_apps_list)
```
yes, that's nicer

orangejenny committed:
Update corehq/apps/users/models.py

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

orangejenny commented:
@millerdev 

> The Save button and radio buttons are a bit crowded, to the point where I initially didn't see the top radio button. Is that typical spacing for those elements? Would it make sense to give the radio buttons a bit more breathing room?

It's using the standard B5 styling for a list of radio buttons (this page was just migrated to B5 in https://github.com/dimagi/commcare-hq/pull/34374). I'm inclined to leave it alone, since we're deprecating this UI.

> Not sure if I'm understanding this correctly. Could it be confusing if, say, a domain configured the new app-specific permissions while also having (pre-existing) conflicting ApplicationAccess restrictions, and had no way to edit the latter as a result of the feature flag being enabled? If yes, would it make sense to ignore ApplicationAccess restrictions if the feature flag is enabled and any new app-specific permissions have been configured?

The new flag enables the old UI (yes this is confusing when reading this PR - I hope it'll be intuitive to future readers who are just looking at the flag, since flags typically enable behavior instead of disabling it). Domains with pre-existing ApplicationAccess restrictions are the ones that are getting the new flag enabled, so that they'll continue to have access. The flag is added so that domains who _don't_ currently use ApplicationAccess, which is most domains, will no longer see it. Does that make sense?

orangejenny committed:
Merge branch 'master' into jls/web-apps-specific-permissions

orangejenny committed:
Removed initial chaeck in get_web_apps_available_to_user

This was just added for https://github.com/dimagi/commcare-hq/pull/34351/,
but it's not needed now that there's a user.can_access_web_app filter further down in the function.

orangejenny committed:
Merge branch 'master' into jls/web-apps-specific-permissions

orangejenny committed:
Merge branch 'master' into jls/web-apps-specific-permissions

orangejenny committed:
Revert "Removed leftover is_commcare_user check"

This reverts commit 27c0cc3fca7cc0fb3d289d6140d77b8b48e78ec6.

orangejenny committed:
Revert "Stopped allowing all mobile users access to web apps"

This reverts commit 16ddf89bf52ae6dbc0f7f756dbc8960bd69f758a.

orangejenny committed:
Added access_web_apps to default mobile worker role

Even though this isn't checked, it's better for it to reflect actual behavior,
which is that mobile workers all get web apps access.

orangejenny commented:
I'm reverting the last few commits, which change the mobile worker default role and remove the `is_commcare_user` checks for web apps access, as I don't believe they're safe to make without substantially more due diligence and also likely a slower rollout that's explicitly communicated to users.

Many mobile workers have no role set. This makes sense, since the default role is a relatively recent addition and we're lazily migrating mobile workers to use it (see https://github.com/dimagi/commcare-hq/pull/32057). For those mobile workers, if you ask for their role, you get the default role back (see https://github.com/dimagi/commcare-hq/pull/32008/), but directly accessing their permissions doesn't go through that default role:

```
> user = CouchUser.get_by_user_id('<a_user_id_with_no_role_set>')

> user.domain_membership.role_id   # no role set directly on the domain membership

> user.domain_membership.role  # access role on object, get back default role, great
UserRole(domain='3156-down', name='Mobile Worker Default')

> user.get_domain_membership('3156-down').role.permissions.access_web_apps  # default role can access web apps, great
True

> user.can_access_web_apps()  # checking permissions on user does not use default role, alas
False
```

That means the migration to add `access_web_apps` to all default mobile worker roles is insufficient to actually grant them access to web apps. We'd need to either do a non-lazy migration to assign roles to all mobile workers, or update permissions checking so that it respects the default role if no role is explicitly set. Both of those are out of scope.

millerdev commented on corehq/apps/users/migrations/0061_grant_web_apps_default_mobile_worker.py:
Does removing this file have any implications for live environments (possibly staging)? Should the change be reversed?

Also, what happens when a future `0061` migration is added to this app? Will it appear to have been applied in any environment where this was applied? Maybe this has not been applied anywhere, in which case there is no concern.

orangejenny replied:
Important questions, and yes, this migration had already been run on staging. I ran `migrate users 0060` on staging to "undo" the migration from django's perspective, and then undid the changes in a shell (running the code that was in the migration but flipping `access_web_apps` to False).

esoergel commented:
@orangejenny that behavior looks very bizarre to me.  I just took a look and I think this is a bug.  Check this out:
```python
>>> user.can_access_web_apps()
False
>>> user.can_access_web_apps(self.domain)
True
```
That's the core of the problem, I believe.  `can_access_web_apps` goes through `__getattr__` and `_get_perm_check_fn`
https://github.com/dimagi/commcare-hq/blob/31ddfc017b694e6b574dfedc9a5a2487ae50a48e/corehq/apps/users/models.py#L1600-L1615
So the two calls above resolve to
```python
>>> user.has_permission(None, 'access_web_apps')
False
>>> user.has_permission(self.domain, 'access_web_apps')
True
```
That is, this will work if `domain` is passed to the `can_` call, or if it is in a request context, where `current_domain` is populated.  This is a gross inconsistency in the code.

esoergel commented:
(I do still support your decision to separate the main changes from addressing this stuff though)

orangejenny committed:
Added  domain to can_access_any_web_apps call

orangejenny committed:
Updated message on legacy page

orangejenny commented:
@esoergel OH. That makes sense. Yeah, my testing was bad. Which is a relief. But yeah, I think I'm going to keep those changes reverted, I'm just gun-shy about sweeping permissions changes.

orangejenny committed:
Updated dropdown option text

orangejenny committed:
Merge branch 'master' into jls/web-apps-specific-permissions

orangejenny committed:
Tagged new flag as USH

USH is responsible for this deprecation.

orangejenny committed:
Updated app-checking to allow for linked apps

orangejenny committed:
Merge branch 'master' into jls/web-apps-specific-permissions

orangejenny committed:
Revert "Updated app-checking to allow for linked apps"

This reverts commit 92099af17579f2624fe4c46745246d183d92317f.

orangejenny committed:
Updated linked domain update to replace app ids with linked app ids

millerdev commented on corehq/apps/linked_domain/updates.py:
```diff
@@ -429,6 +433,12 @@ def update_user_roles(domain_link, is_pull=False, overwrite=False):
 
         permissions = HqPermissions.wrap(upstream_role_def["permissions"])
+        if permissions.web_apps_list:
+            permissions.web_apps_list = [
+                downstream_id for downstream_id, upstream_id in app_id_map.items()
+                if upstream_id in permissions.web_apps_list
+            ]
```
I assume this new usage of app ids does not require any data migrations since this is a new feature that is only deployed on staging where old data either does not exist or does not matter. Is that right?

orangejenny replied:
Yes, exactly.

esoergel commented on corehq/apps/cloudcare/views.py:
```diff
@@ -114,14 +114,22 @@ def get_web_apps_available_to_user(self, domain, user):
+        def _can_access_web_app(user, domain, app):
+            if user.can_access_web_app(domain, app.get('copy_of', app.get('_id'))):
+                return True
+            if user.can_access_web_app(domain, app.get('upstream_app_id')):
```
Can you elaborate on this one?  This is for the situation where there are two apps on the same domain, one linked to the other?  Would these show up as different entries on the roles page where you select which apps are accessible?

esoergel commented on corehq/apps/linked_domain/tests/test_update_roles.py:
```diff
@@ -220,6 +221,25 @@ def test_error_message_indicates_pull(self):
+
+        update_user_roles(self.domain_link)
+
+        role = UserRole.objects.get_by_domain(self.downstream_domain)[0]
+        self.assertListEqual(role.permissions.web_apps_list, [downstream_app.id])
```
Ah this clarifies it, thank you.  In my words, it's that a role on an upstream domain with access to an app needs to be propagated to the downstream domain with the app access updated to the corresponding downstream app.  Did I get that right?

orangejenny replied:
yep, exactly

esoergel reviewed (approved):
Good catch with the linked domains thing, that wasn't on my radar
